sda-commons-server-auth/src/main/java/org/sdase/commons/server/opa/extension/OpaInputExtension.java

```diff
@@ -0,0 +1,84 @@
+package org.sdase.commons.server.opa.extension;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import javax.ws.rs.container.ContainerRequestContext;
+
+/**
+ * An extension to provide additional data to the {@link
+ * org.sdase.commons.server.opa.filter.model.OpaInput} before sending it to the Open Policy Agent.
+ *
+ * <p>Implementing this class should be the <i>last resort</i>. It is more favourable to depend on
+ * the existing {@link org.sdase.commons.server.opa.extension extensions} and instead use the
+ * constraint feature of the {@link
+ * org.sdase.commons.server.opa.OpaJwtPrincipal#getConstraintsAsEntity(Class)} to receive data from
+ * the policy decider and use it to decide based on it in your service.
+ *
+ * <p>Each extension is added in its own namespace, that means that it's data is accessible in a
+ * single sub-property. This contents can be represented by any object that is serializable by an
+ * {@link ObjectMapper}.
+ *
+ * <p>Example that returns a boolean that was serialized with the provided {@link ObjectMapper}:
+ *
+ * <pre>
+ *   {
+ *     "jwt": "…",
+ *     "path": ["…", "…"],
+ *     "httpMethod": "GET",
+ *     "myExtension": true
+ *   }
+ * </pre>
+ *
+ * <p>Example that returns an object that was serialized with the provided {@link ObjectMapper}:
+ *
+ * <pre>
+ *   {
+ *     "jwt": "…",
+ *     "path": ["…", "…"],
+ *     "httpMethod": "GET",
+ *     "myExtension": {
+ *       "myBoolean": true,
+ *       "myString": "asdf",
+ *       "myArray": ["…", "…", "…"]
+ *     }
+ *   }
+ * </pre>
+ *
+ * Please note that is is prohibited to override properties that are already set in the original
+ * {@link org.sdase.commons.server.opa.filter.model.OpaInput}.
+ */
+public interface OpaInputExtension {
+
+  /**
+   * The namespace (i.e. property name) where this extension is included. Please stick to naming the
+   * property in snake_case or camelCase.
+   *
+   * @return the namespace to store the value in.
+   */
+  String getNamespace();
+
+  /**
+   * When registered, it is called in {@link
+   * org.sdase.commons.server.opa.filter.OpaAuthFilter#filter(ContainerRequestContext)}. The return
+   * value is added as child of the property name defined by {@link #getNamespace()}.
+   *
+   * <p>Example that adds the property {@code "myExtension": true}:
+   *
+   * <pre>
+   *   &#64;Override
+   *   public String getNamespace() {
+   *     return "myExtension";
+   *   }
+   *
+   *   &#64;Override
+   *   public JsonNode filter(ContainerRequestContext requestContext, ObjectMapper mapper) {
+   *     return mapper.valueToTree(true);
+   *   }
+   * </pre>
+   *
+   * @param requestContext the request context
+   * @param mapper the object mapper to serialize values
+   * @return the JsonNode that should be added as child of the extension's namespace property.
+   */
+  JsonNode filter(ContainerRequestContext requestContext, ObjectMapper mapper);
```

Comment:
I think the implementing class should not be aware about the technical detail that this implementation will be used in a filter. In my opinion we should find a name that clarifies what should happen in this method, maybe something like `createAdditionalPropertyContent(…)`.

I also think, the mapper should not be part of the method signature. The class can be instantiated with an appropriate mapper.

Have you considered returning either `Object` or a generic type? The interface may be implemented easier in that case. Is there a technical reason, that prevents merging the objects later?

Reply:
I first used `Object` but found it too unprecise and hard to understand what this means without knowing the documentation. But I haven't though about a generic type.

Reply:
Well, `JsonNode` is as unprecise as `Object` but more complicate to handle.
The only thing you guarantee is that the returned object is serializable by the `ObjectMapper`. But I think that's not worth the complexity of usage.

Reply:
I changed it to a generic type and renamed the method.